### PR TITLE
replace packages.txt with calculated graph

### PR DESCRIPTION
**Pull request by deitch**
This replaces `packages.txt` entirely with `wolfictl text` to generate the graph.

This is a draft for a few reasons:

* I want several pairs of eyes on this, to make sure we get it right
* the graph generation currently runs about 4.5-5 mins; I want to profile it to see where it is taking so much time and fix it up. UPDATE: this is due ti a change in the sorting algorithm. I opened an issue on it, will follow up with it. See [here](https://github.com/dominikbraun/graph/issues/129) - UPDATE: with a few small caching improvements, now down to 3.5 mins. Not enough, but getting there
* ~there remains an area in CI where we use `packages.txt`, but it is the last step in a job, so I don't understand how it is used, and thus how to replace it~

To reproduce the work this will do:

1. Clone this repo
2. Make sure you have a recent version of `wolfictl` or use the container image for `wolfictl` or `sdk`
3. Run `wolfictl text --dir . --type name` and wait a few minutes for the output

### CODE REVIEW

**Comment by deitch** [comment]
> the graph generation currently runs about 4.5-5 mins; I want to profile it to see where it is taking so much time and fix it up 

I ran a profile, the overwhelming majority of the time is in the graph sort. Attached. I will remove the stable sort and see if that is the issue, either way, I will report it upwards.

flame-graph-stabel-sort

**Comment by deitch** [comment]
`wolfictl lint` still checks the Makefile. I could have removed that check from `wolfictl lint`, but then the current lint wouldn't work, and other places that use it would have had issues. Instead, I updated the call to `--skip-rule no-makefile-entry-for-package`

**Comment by kaniini** [comment]
Build order is almost right, but not quite yet.  `build-base`, which explicitly depends on `glibc-dev`, comes before `glibc`.  But `gcc` is now coming before `build-base`, so we are much closer.

**Comment by deitch** [comment]
There was an issue. It was not allowing self-fulfilling dependencies, and checking in an incorrect way. There is an open PR that fixes that issue.

However, it opens several new issues with the actual `wolfi-dev/os` dependencies. Will post here.

**Comment by deitch** [comment]
```time="2023-06-18T12:15:38+03:00" level=error msg="unresolvable cycle: wolfi-baselayout:20230201-r3@local -> glibc-locale-posix:2.37-r7@local, caused by: glibc-locale-posix:2.37-r7@local -> glibc:2.37-r7@local -> wolfi-baselayout:20230201-r3@local"
time="2023-06-18T12:15:38+03:00" level=error msg="unresolvable cycle: pax-utils:1.3.7-r0@local -> meson:1.1.1-r0@local, caused by: meson:1.1.1-r0@local -> scanelf:1.3.7-r0@local -> pax-utils:1.3.7-r0@local"
time="2023-06-18T12:15:38+03:00" level=error msg="unresolvable cycle: pcre2:10.42-r2@local -> binutils:2.40-r3@local, caused by: binutils:2.40-r3@local -> wolfi-baselayout:20230201-r3@local -> glibc-locale-posix:2.37-r7@local -> glibc:2.37-r7@local -> grep:3.11-r0@local -> pcre2-dev:10.42-r2@local -> pcre2:10.42-r2@local"
time="2023-06-18T12:15:39+03:00" level=error msg="unresolvable cycle: rustls-ffi:0.10.0-r0@local -> git:2.41.0-r0@local, caused by: git:2.41.0-r0@local -> curl-dev:8.1.2-r0@local -> curl:8.1.2-r0@local -> rustls-ffi:0.10.0-r0@local"
```

There are 4 dependencies here, but they fall into 2 categories.

### wolfi-baselayout

1. `wolfi-baselayout` [depends on](https://github.com/wolfi-dev/os/blob/main/wolfi-baselayout.yaml#L10) `glibc-local-posix`
2. which is a [subpackage of glibc](https://github.com/wolfi-dev/os/blob/main/glibc.yaml#L457-L462)
3. `glibc`, in turn, depends [runtime on wolfi-baselayout](https://github.com/wolfi-dev/os/blob/main/glibc.yaml#L10).

This is an unresolvable loop.

### pax-utils

pax-utils is due to pipelines

1.  [pax-utils depends in its pipelines on `meson`](https://github.com/wolfi-dev/os/blob/main/pax-utils.yaml#L31)
7. which in its pipelines [uses `strip`](https://github.com/wolfi-dev/os/blob/main/meson.yaml#L32)
8. whose [needs include `scanelf`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/strip.yaml#L6)
9. which in turn is a [subpackage of `pax-utils`](https://github.com/wolfi-dev/os/blob/main/pax-utils.yaml#L41)

And we have a cycle.

We can do the same thing with the others:

1. `pcre2:10.42-r2@local` [depends on `strip`](https://github.com/wolfi-dev/os/blob/main/pcre2.yaml#L33)
2. which [depends on `binutils`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/strip.yaml#LL5C1-L5C1)
3. which depends on `wolfi-baselayout`
4. which depends on `glibc-locale-posix`
5. which is a subpackage of `glibc`
6. which depends on `grep`
7. which depends on `pcre2-dev`
8. which is provided by `pcre`

### rustls-ffi

1. `rustls-ffi:0.10.0-r0@local` depends on the pipeline [git-checkout](https://github.com/wolfi-dev/os/blob/main/rustls-ffi.yaml#L21)
2. which depends on `git`
3. which depends on `curl-dev`
4. which is a [subpackage of `curl`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L64)
5. which [depends on `rustls-ffi`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L20)

And we have a cycle

**Comment by kaniini** [comment]
I think we can drop the glibc dependency on wolfi-baselayout.

Meson, being python, can just drop `strip`.

Not sure what to do about `rustls-ffi`.

**Comment by deitch** [comment]
I can open PRs for the first two. As you said, not sure what to do about the rust one. 

What about pcre?

**Comment by deitch** [comment]
> I think we can drop the glibc dependency on wolfi-baselayout.

See #2937 

> Meson, being python, can just drop strip.

See #2938

It only partially solves the job. It frees up one loop, but reveals another.

1. pax-utils depends buildtime on  meson
2. which depends buildtime on busybox
3. which depends on strip
4. which depends on scanelf
5. which is a subpackage of pax-utils

**Comment by deitch** [comment]
With the following issues resolved - glibc/wolfi-baselayout, meson/strip - here is what is left.

### pax-utils

pax-utils:1.3.7-r0@local -> meson:1.1.1-r0@local
, caused by: meson:1.1.1-r0@local -> busybox:1.36.1-r0@local -> scanelf:1.3.7-r0@local -> pax-utils:1.3.7-r0@local"


1.  [pax-utils depends in its pipelines on `meson`](https://github.com/wolfi-dev/os/blob/main/pax-utils.yaml#L31)
2. which depends in [its environment on `busybox`](https://github.com/wolfi-dev/os/blob/main/meson.yaml#L16)
3. which depends in [its pipelines on `strip`](https://github.com/wolfi-dev/os/blob/main/busybox.yaml#L70)
4. whose [needs include `scanelf`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/strip.yaml#L6)
5. which in turn is a [subpackage of `pax-utils`](https://github.com/wolfi-dev/os/blob/main/pax-utils.yaml#L41)

### rustls-ffi

1. `rustls-ffi:0.10.0-r0@local` depends on the pipeline [git-checkout](https://github.com/wolfi-dev/os/blob/main/rustls-ffi.yaml#L21)
2. which depends on `git`
3. which depends on `curl-dev`
4. which is a [subpackage of `curl`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L64)
6. which [depends on `rustls-ffi`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L20)

### pcre2

1. `pcre2:10.42-r2@local` [depends on `strip`](https://github.com/wolfi-dev/os/blob/main/pcre2.yaml#L33)
2. which [depends on `binutils`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/strip.yaml#LL5C1-L5C1)
3. which depends on `wolfi-baselayout`
4. which depends on `glibc-locale-posix`
5. which is a subpackage of `glibc`
7. which depends on `grep`
8. which depends on `pcre2-dev`
9. which is provided by `pcre`

**Comment by deitch** [comment]
With #3026 in place, the following is the updated list:

### pax-utils

1.  `pax-utils` [depends in its pipelines on `meson`](https://github.com/wolfi-dev/os/blob/main/pax-utils.yaml#L31)
2. which depends in [its environment on `ca-certificates-bundle`](https://github.com/wolfi-dev/os/blob/main/meson.yaml#L17)
3. which is [a subpackage of `ca-certificates`](https://github.com/wolfi-dev/os/blob/main/ca-certificates.yaml#L63)
4. which depends in [its environment on `openssl-dev`](https://github.com/wolfi-dev/os/blob/main/ca-certificates.yaml#L22)
5. which is [a subpackage of `openssl`](https://github.com/wolfi-dev/os/blob/main/openssl.yaml#L123)
6. which [depends in its pipelines on `strip`](https://github.com/wolfi-dev/os/blob/main/openssl.yaml#L75)
7. whose [needs include `scanelf`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/strip.yaml#L6)
8. which in turn is a [subpackage of `pax-utils`](https://github.com/wolfi-dev/os/blob/main/pax-utils.yaml#L41)

### rustls-ffi

1. `rustls-ffi` depends on the pipeline [git-checkout](https://github.com/wolfi-dev/os/blob/main/rustls-ffi.yaml#L21)
2. which [depends on `git`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/git-checkout.yaml#L5)
3. which [depends on `curl-dev`](https://github.com/wolfi-dev/os/blob/main/git.yaml#L16)
4. which is a [subpackage of `curl`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L64)
9. which [depends on `rustls-ffi`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L20)

### pcre2

1. `pcre2` [depends in its environment on `build-base`](https://github.com/wolfi-dev/os/blob/main/pcre2.yaml#L15)
2. which [depends on `glibc-dev`](https://github.com/wolfi-dev/os/blob/main/build-base.yaml#L15)
3. which [is a subpackage of `glibc`](https://github.com/wolfi-dev/os/blob/main/glibc.yaml#L331)
4. which [depends in its environment on `grep`](https://github.com/wolfi-dev/os/blob/main/glibc.yaml#L40)
5. which [depends on `pcre2-dev`](https://github.com/wolfi-dev/os/blob/main/grep.yaml#L21)
6. which [is a subpackage of `pcre`](https://github.com/wolfi-dev/os/blob/main/pcre2.yaml#L44)

**Comment by deitch** [comment]
I think I have figured out some of this.

Let's use the last one, `pcre2`, as an example.

When we try to resolve `pcre2 -> build-base`, the resolver includes one the local packages, because that is what `pcre2` has in its `environment`. We use that for the whole stack and for resolving cycles.

However, that actually is not correct. It is correct for the specific `pcre2 -> build-base`. But `build-base` might have a different resolver. It doesn't, as it is in `runtime` (which always is local only). If we keep going down the stack, we eventually reach stage 4, where `glibc -> grep`.  `glibc`'s definition allows it to come from local or upstream (bootstrap). So this actually is resolvable. That would make the cycle go away.

Unfortunately, it means rebuilding data about the graph, but hopefully not too hard. A few days at least.

**Comment by deitch** [comment]
After more work handling the above, down to the following loops.

Original logs, followed by detailed explanation.

```
time="2023-06-26T20:31:34-07:00" level=error msg="unresolvable cycle: py3.11-setuptools:68.0.0-r0@local -> scanelf:1.3.7-r0@local, caused by: scanelf:1.3.7-r0@local -> pax-utils:1.3.7-r0@local -> meson:1.1.1-r0@local -> py3.11-setuptools:68.0.0-r0@local"
time="2023-06-26T20:31:34-07:00" level=error msg="unresolvable cycle: rustls-ffi:0.10.0-r0@local -> git:2.41.0-r0@local, caused by: git:2.41.0-r0@local -> curl-dev:8.1.2-r0@local -> curl:8.1.2-r0@local -> rustls-ffi:0.10.0-r0@local"
time="2023-06-26T20:31:34-07:00" level=error msg="unresolvable cycle: rustls-ffi:0.10.0-r0@local -> libLLVM-15:15.0.7-r1@local, caused by: libLLVM-15:15.0.7-r1@local -> llvm15:15.0.7-r1@local -> cmake:3.24.2-r2@local -> curl-dev:8.1.2-r0@local -> curl:8.1.2-r0@local -> rustls-ffi:0.10.0-r0@local"
time="2023-06-26T20:31:34-07:00" level=error msg="unresolvable cycle: samurai:1.2-r2@local -> scanelf:1.3.7-r0@local, caused by: scanelf:1.3.7-r0@local -> pax-utils:1.3.7-r0@local -> meson:1.1.1-r0@local -> samurai:1.2-r2@local"
```

### rustls-ffi (first loop)

1. `rustls-ffi` depends on the pipeline [git-checkout](https://github.com/wolfi-dev/os/blob/main/rustls-ffi.yaml#L21)
2. which [depends in its needs on `git`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/git-checkout.yaml#L5)
3. which [depends on its environment `curl-dev`](https://github.com/wolfi-dev/os/blob/main/git.yaml#L16)
4. which is a [subpackage of `curl`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L64)
5. which [depends in its environment on `rustls-ffi`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L20)

### rustls-ffi (second loop)

1. `rustls-ffi` [depends in its environment on libLLVM-15](https://github.com/wolfi-dev/os/blob/main/rustls-ffi.yaml#L13)
2. which [is a subpackage of `llvm-15`](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/llvm15.yaml#L112)
3. which [depends in its environment on `cmake`](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/llvm15.yaml#L16)
4. which [depends in its environment on `curl-dev`](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/cmake.yaml#L17)
4. which is a [subpackage of `curl`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L64)
5. which [depends in its environment on `rustls-ffi`](https://github.com/wolfi-dev/os/blob/main/curl.yaml#L20)

NOTE: both rustls-ffi loops in the end depend on `curl-dev` -> `curl` -> `rustls-ffi`

### samurai

1. `samurai` [uses `strip`](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/samurai.yaml#L40)
2. which [depends in its needs on `scanelf`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/strip.yaml#L6)
3. which [is a subpackage of `pax-utils`](https://github.com/wolfi-dev/os/blob/main/pax-utils.yaml#L41)
4. which [uses various `meson` in its pipeline](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/pax-utils.yaml#L31-L35)
5. which [needs `meson`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/meson/compile.yaml#L5)
6. which [depends in its environment on `samurai`](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/meson.yaml#L20)

### py3.11-setuptools

1. `py3.11-setuptools` [uses in its pipelines `strip`](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/py3.11-setuptools.yaml#L35)
2. which [depends in its needs on `scanelf`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/strip.yaml#L6)
3. which [is a subpackage of `pax-utils`](https://github.com/wolfi-dev/os/blob/main/pax-utils.yaml#L41)
4. which [uses various `meson` in its pipeline](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/pax-utils.yaml#L31-L35)
5. which [needs `meson`](https://github.com/chainguard-dev/melange/blob/12a1c4076cc005f04798f710c2fce3ed6cb47904/pkg/build/pipelines/meson/compile.yaml#L5)
6. which [depends in its environment on `py3-setuptools `](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/meson.yaml#L18)
7. which [is provided by `py3.11-setuptools`](https://github.com/wolfi-dev/os/blob/4aa584814c34cde4085a0000129a78c089662176/py3.11-setuptools.yaml#L10)

**Comment by kaniini** [comment]
py3.11-setuptools (and the other related packages) can just drop the `strip` pipeline.

**Comment by kaniini** [comment]
To clear the samurai loop, just run strip on `${{targets.destdir}}/usr/bin/samu` directly.

**Comment by kaniini** [comment]
For `curl-dev`, we need a `curl-bootstrap-dev` which provides `curl-dev` and does not depend on rust.  We prioritize it lower than normal `curl`.

**Comment by deitch** [comment]
Handled the first two; will look more deeply into the others tomorrow.

**Comment by deitch** [comment]
> For curl-dev, we need a curl-bootstrap-dev which provides curl-dev and does not depend on rust. We prioritize it lower than normal curl.

It is slightly more complicated, but it makes sense. `curl-dev` is a subpackage of `curl`, but it also depends on `libcurl4`, which is provided by two subpackages in `curl`: `libcurl-openssl4` and `libcurl-rustls4` (simplified below).

```yaml
subpackages:
  - name: "curl-dev"
    dependencies:
      runtime:
        - libcurl4

  - name: "libcurl-openssl4"
    description: "curl library (openssl backend)"
    dependencies:
      provides:
        - libcurl4=7.87.1
      provider-priority: 5

  - name: "libcurl-rustls4"
    description: "curl library (rustls backend)"
    dependencies:
      provides:
        - libcurl4=7.87.1
      provider-priority: 10
```

> For curl-dev, we need a curl-bootstrap-dev which provides curl-dev and does not depend on rust

Would that be pretty much identical to `curl`, just a part of it, i.e. no building `libcurl-rustls4`?

This whole thing is a bit off. `libcurl-rustls4` is a sub of `curl4` (which makes sense), is depended upon by `curl-dev` (which makes sense), and depends upon `rustls-ffi` (which makes sense). `rustls-ffi` is a direct dependency of `libcurl-rustls4`, so how can it depend on `git` which depends upon `curl-dev` which depends upon `rustls-ffi`?

The others were because we consciously added `strip`, so we can consciously remove them. These are core to the environment or runtime.

You know what else would work? If these could use `curl-dev` from bootstrap. But it doesn't exist there.

But partially I am suggesting it because I don't fully get what the `libcurl-bootstrap-dev` is. 😁

**Comment by deitch** [comment]
Once https://github.com/wolfi-dev/wolfictl/pull/285 is into wolfictl, this will be ready. Attached is the latest generated `packages.txt`.
[packages.txt](https://github.com/wolfi-dev/os/files/11940180/packages.txt)

**Comment by deitch** [comment]
Do **not** merge this in. It is good to approve, I am going to wait until I see that the latest version of `wolfi-dev/sdk` is released on the nightly cycle and then I will merge it in myself.

**Comment by jdolitsky** [comment]
> Do not merge this in. It is good to approve, I am going to wait until I see that the latest version of wolfi-dev/sdk is released on the nightly cycle and then I will merge it in myself.

FWIW we can manually trigger this build this whenever needed

**Comment by deitch** [comment]
Yeah, but I missed something on the Makefile anyways. Working that now

**Comment by dlorenc** [comment]
Didn't the symlinked directories break arm builds last time? Did that get fixed?

**Comment by deitch** [comment]
> Didn't the symlinked directories break arm builds last time? Did that get fixed?

I am pretty sure they did, but I cannot remember what it was that fixed it. As soon as I have the Makefile demangled, I will track it down.

**Comment by deitch** [comment]
Some updates here:

1. I fixed the Makefile issue. Actually, at @kaniini 's prodding, I removed the complex `.packagerules` generation, and replaced it with some simple targets. However, to do so, I had to change `make packages/openssl` with `make package/openssl` or makefile got confused. I tracked down where we use this in GHA workflows, but another pair of eyes (or two) on this would be helpful.
2. That makefile bit requires an extra call, which slows it down a bit. I don't think it is slower than currently (likely similar), but another pair of eyes, again, would be helpful.
3. The issue with symlinks was `kontext`, the revert and comment are [here](https://github.com/wolfi-dev/os/pull/2129).

Those need review, then we wait for wolfictl to be updated in wolfi/sdk by tomorrow.

**Comment by kaniini** [comment]
Symlinks should be good in kontext now

**Comment by deitch** [comment]
1. Updated to `ghcr.io/wolfi-dev/sdk:latest@sha256:fd8c71214f6455c75ec44ae99eb9f7ffc85f260ccce69d4367eb2e5d568facd9`. Checked the version from there to be:
```
GitVersion:    6a0feae
GitCommit:     6a0feae8fcfd8771f03575bdcc8158a5cb7d01a6
```
which is what is in latest build on this branch, so it is correct.
2. @kaniini confirmed above that `kontext` is fine with symlinks
3. Just need agreement on the Makefile changes and this can go in.

I will remove the unused `packages.txt` in a later PR.

**Comment by joshrwolf** [comment]
this shouldn't block, but I get these errors when trying to reproduced this:

```bash
➜ wolfictl text --dir . --type name
Error: unable to build graph:
envoy-1.26.2-r0: unable to resolve dependency gcc=12.2.0-r11: could not find package gcc=12.2.0-r11 in indexes
envoy-1.26.2-r0: unable to resolve dependency libgcc=12.2.0-r11: could not find package libgcc=12.2.0-r11 in indexes
envoy-1.26.2-r0: unable to resolve dependency libstdc++-dev=12.2.0-r11: could not find package libstdc++-dev=12.2.0-r11 in indexes
envoy-1.26.2-r0: unable to resolve dependency libstdc++=12.2.0-r11: could not find package libstdc++=12.2.0-r11 in indexes
FATA[0000] error during command execution: unable to build graph:
envoy-1.26.2-r0: unable to resolve dependency gcc=12.2.0-r11: could not find package gcc=12.2.0-r11 in indexes
envoy-1.26.2-r0: unable to resolve dependency libgcc=12.2.0-r11: could not find package libgcc=12.2.0-r11 in indexes
envoy-1.26.2-r0: unable to resolve dependency libstdc++-dev=12.2.0-r11: could not find package libstdc++-dev=12.2.0-r11 in indexes
envoy-1.26.2-r0: unable to resolve dependency libstdc++=12.2.0-r11: could not find package libstdc++=12.2.0-r11 in indexes
```

digging in, I updated `envoy` to use:

```yaml
# OLD
      - gcc=12.2.0-r11
      - libstdc++=12.2.0-r11
      - libstdc++-dev=12.2.0-r11
      - libgcc=12.2.0-r11
# NEW
      - gcc-12
      - libstdc++-12
      - libstdc++-12-dev
      - libgcc # This should be `libgcc-12`
```

and that got past that error, but flagged a new one:

```yaml
❯ wolfictl text --dir . --type name
Error: unable to build graph:
rabbitmq-server-3.12.0-r1: unable to resolve dependency elixir<1.15: could not find package elixir<1.15 in indexes
FATA[0000] error during command execution: unable to build graph:
rabbitmq-server-3.12.0-r1: unable to resolve dependency elixir<1.15: could not find package elixir<1.15 in indexes
```

which I only got to work by setting to the latest `elixir`, which in theory breaks the `rabbitmq-server` package.

am I doing something wrong? or is `wolfictl` not parsing these edge cases correctly?

**Comment by deitch** [comment]
This appears to be correct. [gcc-12.yaml](https://github.com/wolfi-dev/os/blob/main/gcc-12.yaml#L1-L4) has version 12.3.0-r0:

```yaml
package:
  name: gcc-12
  version: 12.3.0
  epoch: 0
```

But [envoy.yaml requires](https://github.com/wolfi-dev/os/blob/main/envoy.yaml#L17-L20) _exactly_ 12.2.0-r11:

```yaml
      - gcc=12.2.0-r11
      - libstdc++=12.2.0-r11
      - libstdc++-dev=12.2.0-r11
      - libgcc=12.2.0-r11
```

There is a similar issue with [rabbitmq depending on `elixir<1.15`](https://github.com/wolfi-dev/os/blob/main/rabbitmq-server.yaml#L20), when [elixir.yaml is](https://github.com/wolfi-dev/os/blob/main/elixir.yaml#L3-L4) 1.15.2-r0.

**Comment by deitch** [comment]
You could make the argument that, "we had those packages at one time, and we loaded them up into our registry, so maybe they already are there." That is a good argument, but it loosens the constraints. The graph checks if we can build it with the most constrained situation possible, which means without the cache at all (from scratch). These changes prevent it.

File: Makefile [hunk 1]
```diff
@@ -48,31 +46,24 @@ clean:
...
+	$(eval pkgver := $(shell $(MELANGE) package-version $(yamlfile)))
+	$(MAKE) yamlfile=$(yamlfile) pkgname=$* packages/$(ARCH)/$(pkgver).apk
+
+packages/$(ARCH)/%.apk: $(KEY)
+	@mkdir -p ./$(pkgname)/
+	$(eval SOURCE_DATE_EPOCH ?= $(shell git log -1 --pretty=%ct --follow $(yamlfile)))
+	@SOURCE_DATE_EPOCH=$(SOURCE_DATE_EPOCH) $(MELANGE) build $(yamlfile) $(MELANGE_OPTS) --source-dir ./$(pkgname)/ --log-policy builtin:stderr,$(TARGETDIR)/buildlogs/$*.log"
 
 dev-container:
 	docker run --privileged --rm -it -v "${PWD}:${PWD}" -w "${PWD}" ghcr.io/wolfi-dev/sdk:latest@sha256:3ef78225a85ab45f46faac66603c9da2877489deb643174ba1e42d8cbf0e0644
```
Comment (joshrwolf) [inline Makefile:69]: happy to do this in a different PR if you want, but somehow this `dev-container` got out of date enough to not include `wolfictl`, which, with this PR, is now needed for the dev loop

```bash
❯ crane export ghcr.io/wolfi-dev/sdk:latest@sha256:3ef78225a85ab45f46faac66603c9da2877489deb643174ba1e42d8cbf0e0644 | tar tvf - | grep -i wolfictl

```

vs:

```bash
❯ crane export ghcr.io/wolfi-dev/sdk:latest@sha256:3ba6e392eff7f09493c62b8a6bff4b9378ecccc27e5dc4ba0fa9f2a0e95c666f | tar tvf - | grep -i wolfictl
-rwxr-xr-x  0 root   root 72379332 Dec 31  1969 usr/bin/wolfictl
                                                                                                                                                               
```

```suggestion
	docker run --privileged --rm -it -v "${PWD}:${PWD}" -w "${PWD}" ghcr.io/wolfi-dev/sdk:latest@sha256:3ba6e392eff7f09493c62b8a6bff4b9378ecccc27e5dc4ba0fa9f2a0e95c666f
```

Comment (deitch) [reply Makefile:69]: It probably is because we have digestabot update these files directly. If we had a single file with all of them, and the Makefile and workflows just pulled them in, it all would get updated automatically. Maybe I will open a PR for it.

In the meantime, I will rebase and update it.

File: Makefile [hunk 2]
```diff
@@ -48,31 +46,24 @@ clean:
...
--include .packagerules
+package/%:
+	$(eval yamlfile := $*.yaml)
+	$(eval pkgver := $(shell $(MELANGE) package-version $(yamlfile)))
+	$(MAKE) yamlfile=$(yamlfile) pkgname=$* packages/$(ARCH)/$(pkgver).apk
+
+packages/$(ARCH)/%.apk: $(KEY)
+	@mkdir -p ./$(pkgname)/
+	$(eval SOURCE_DATE_EPOCH ?= $(shell git log -1 --pretty=%ct --follow $(yamlfile)))
+	@SOURCE_DATE_EPOCH=$(SOURCE_DATE_EPOCH) $(MELANGE) build $(yamlfile) $(MELANGE_OPTS) --source-dir ./$(pkgname)/ --log-policy builtin:stderr,$(TARGETDIR)/buildlogs/$*.log"
```
[the remark this answers is not in the document]
Comment (deitch) [reply Makefile]: I was wondering why I had that. Nice catch. Fixed.

**Comment by kaniini** [comment]
> These changes prevent it.

Yes, we should not allow build-depending on something that happens to be in the cache.  If we want to pin to an older dependency, then we should do the work to properly set that dependency up to be slotted.

**Review by kaniini — APPROVED** [review]
